clippy_lints/src/needless_deref.rs

```diff
@@ -0,0 +1,84 @@
+use clippy_utils::diagnostics::span_lint_and_help;
+use clippy_utils::source::snippet_opt;
+use rustc_hir::{ExprKind, UnOp};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::mir::Mutability;
+use rustc_middle::ty;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for `&*(&T)`.
+    ///
+    /// ### Why is this bad?
+    /// When people deref on an immutable reference `&T`, they may expect return `&U`.
+    /// Accutually `&* (&T)` is still `&T`.
+    /// if you want to deref explicitly, `&** (&T)` is what you need.
+    /// If you want to reborrow, `&T` is enough (`&T` is Copy).
+    ///
+    /// ### Known problems
+    /// None.
+    ///
+    /// ### Example
+    /// ```rust
+    /// let s = &String::new();
+    ///
+    /// // Bad
+    /// let a: &String = &* s;
+    /// foo(&*s);
+    ///
+    /// // Good
+    /// let a: &String = s;
+    /// foo(&**s);
+    ///
+    /// fn foo(_: &str){ }
+    /// ```
+    pub NEEDLESS_DEREF,
+    complexity,
+    "deref on an immutable reference returns the same type as itself"
+}
+
+declare_lint_pass!(NeedlessDeref => [NEEDLESS_DEREF]);
+
+impl LateLintPass<'_> for NeedlessDeref {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, e: &'tcx rustc_hir::Expr<'_>) {
+        if_chain! {
+            if !e.span.from_expansion();
+            if let ExprKind::AddrOf(_, Mutability::Not, addrof_target) = e.kind;
+            if let ExprKind::Unary(UnOp::Deref, deref_target) = addrof_target.kind;
+            if !matches!(deref_target.kind, ExprKind::Unary(UnOp::Deref, ..) );
+            let inner_ty = cx.typeck_results().expr_ty(deref_target);
+            if let ty::Ref(_, _, Mutability::Not) = inner_ty.kind();
+            then{
+
+                let map = cx.tcx.hir();
+                if let Some(parent_hir_id) = map.find_parent_node(e.hir_id){
+                    let span=map.span(parent_hir_id);
+                    if span.from_expansion() {
+                        return;
+                    }
+                    let parent_node = map.find(parent_hir_id);
+                    if let Some(rustc_hir::Node::Expr(parent_expr)) = parent_node {
+                        if  matches!(parent_expr.kind, ExprKind::Unary(UnOp::Deref, ..)) ||
+                            matches!(parent_expr.kind, ExprKind::AddrOf(..) ){
+                                return;
+                            }
+                    }
```

Comment:
Why is this necessary?

Reply:
`AddrOf` is explained by previous example

Reply:
## `Unary` :
`* &* (&T)`  will trigger `deref_addrof`.
I think we should deal with `deref_addrof` first.

Reply:
I'd just lint twice. Noisy code = noisy clippy.

Reply:
The suggestion for both lints would be the same. `deref_addrof` will suggest removing the first `*&`, this lint will suggest removing the `&*` part. Both of those suggestions will result in `*&T`. Since it's easy to detect I would say it's reasonable to silence one of them.

A check if `deref_addrof` is allowed on the outer deref is needed though. Use `is_lint_allowed`.

Reply:
> A check if deref_addrof is allowed on the outer deref is needed though. Use is_lint_allowed.

Done.